grpc-client-spring-boot-autoconfigure/src/main/java/net/devh/boot/grpc/client/channelfactory/AbstractChannelFactory.java

```
@@ -132,10 +134,43 @@ protected ManagedChannel newManagedChannel(final String name) {
         final T builder = newChannelBuilder(name);
         configure(builder, name);
         final ManagedChannel channel = builder.build();
+        if (properties.getChannel(name).isImmediateConnect()) {
+            connectOnStartup(name, channel);
+        }
         watchConnectivityState(name, channel);
         return channel;
     }
 
+    private void connectOnStartup(String name, ManagedChannel channel) {
+        final ConnectivityState state = channel.getState(true);
+        final CountDownLatch timeout = new CountDownLatch(1);
+        final AtomicBoolean connected = new AtomicBoolean(false);
+
+        channel.notifyWhenStateChanged(state, () -> {
+            ConnectivityState state1 = channel.getState(false);
+            log.info("First state change {}", state1);
+            if (state1 == ConnectivityState.CONNECTING) {
+                channel.notifyWhenStateChanged(state1, () -> {
+                    ConnectivityState state2 = channel.getState(false);
+                    log.info("Second state change {}", state2);
+                    connected.set(state2 == ConnectivityState.READY);
+                    timeout.countDown();
+                });
+            }
+        });
```

How about only checking for `READY` (regardless of the current state)?
Then the timeout would work even if there are temporary failures.

---

If we check for `READY` state only then in case of connection failure we'll have to wait for a full timeout.
If we see that state changed from `CONNECTING` to `TRANSIENT_FAILURE` we may fail faster and user won't have to wait too much.
So except for code beauty first the original approach will be preferable.

---

OK, then we'll start with this. If feedback requests to change it we might touch it again.